### PR TITLE
Throttle the image request queue while the map is moving to improve performance

### PR DESCRIPTION
**Pull request by zhangyiatmicrosoft**
## Launch Checklist

This is the same as https://github.com/maplibre/maplibre-gl-js/pull/1856 which will be abandoned by AdamS108.
Need transfer the ownership over to me.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by zhangyiatmicrosoft** [comment]
@HarelM since #1856 started way before #2038, it is fair to merge this one first. Please let me know what needs to be done. Thanks

**Comment by HarelM** [comment]
@smellyshovel @zhangyiatmicrosoft is there a good way to solve this conflict?
Will the split of the #2038 PR allow this to be merged in between the small PRs?
Generally speaking, both PRs are in final implantation stages, but both need additional PRs to be fully complete (moving the queue to a different file in this PR and the split to smaller PRs in the other)
If you guys want to review each other's code that can be helpful too.

**Comment by smellyshovel** [comment]
> @smellyshovel @zhangyiatmicrosoft is there a good way to solve this conflict? Will the split of the #2038 PR allow this to be merged in between the small PRs? Generally speaking, both PRs are in final implantation stages, but both need additional PRs to be fully complete (moving the queue to a different file in this PR and the split to smaller PRs in the other) If you guys want to review each other's code that can be helpful too.

Merge this one as it's ready. I think it's gonna be a long time until get to `getImage` in my PRs, so by the time I actually do, I will simply try to include the changes from here into my PR.

File: src/util/ajax.ts [hunk 1]
```diff
@@ -324,81 +324,249 @@ function arrayBufferToCanvasImageSource(data: ArrayBuffer, callback: Callback<Ca
...
+        const theQueue = this;
         const queued = {
             requestParameters,
             callback,
             cancelled: false,
-            cancel() { this.cancelled = true; }
+            completed: false,
+            theQueue,
+            cancel() {
+                this.cancelled = true;
```
[the remark this answers is not in the document]
Comment (drwestco) [reply src/util/ajax.ts:417]: Can just use an arrow function here to capture `this`? This likely crept in if this started out as a patch to the compiled JS.

[the remark this answers is not in the document]
Comment (HarelM) [reply src/util/ajax.ts:417]: No, there are two types of "this" here - one for the cancelled field and the other for the queue. It needs a bit more work...

**Comment by zhangyiatmicrosoft** [comment]
@HarelM consolidating all comments from the previous PR to keep us on the same page:

1) this.throttleControlCallbackHandleCounter++;
Q: Can this reach a number too big to become negative?
A: I don't think so. The design of this counter is to keep track of downstream consumers of ImageRequestQueue per session. It is practically impossible for the number to go beyond 2B.
HOWEVER, I did notice a bug when it comes to removeThrottleControlCallback, in that it was using splice and changed the index.
Will be fixed in next iteration.

2) Two places need strong type: will fix

3) "queue item knows about the queue"
I agree this is wrong and will fix.

4) Regarding the confusing usage of "this":
When rewind a little bit, I see class ImageRequestQueue is (and will always) be used as singleton and start to doubt why "class" is needed to begin with.

It was created based on your comment at Dev 22: "This file has all kind of variables laying around, any chance you can turn it into a class?"
While class nicely wraps things together, it introduces keyword "this" problem and does not minimize well.

Therefore, my proposal is change to "namespace ImageRequestQueue", removing the class.
Still wraps things nicely, with smaller runtime script, and avoid "this" keyword. 

What do you think?

**Comment by zhangyiatmicrosoft** [comment]
New iteration, let’s start from the previous toipics
1.	fixed
2.	a new type “ImageRequestQueueItem” 
3.	"queue item knows about the queue"
The original code was somewhat in this style. The queue is self-driven, meaning as soon as one requested item is processed, it will move on. For this reason it auto advance previously at ajas.ts, line 328 (advanceImageRequestQueue).
It is hard to see and honestly the original code was really difficult to understand as well. <br/>
In AdamS108’s iteration both Queue and Item are classes so it looked bad. 
The real queue data structure is nothing more than an array. No separate class is needed. <br/>
In the latest iteration, “namespace ImageRequest”” wraps everything, and it should be much easier to read.
4.	Fixed

More optimizations:
-	Finally moved to a separate file, so won’t have to deal with merge conflicts on ajax.ts. Everything gets moved around so code diff did not help much anyway.
-	Moved a few definitions around as well.
-	Generic code clean ups and shortened a few function/field names in favor of smaller scripts. 
-	Happy to say because of all the above, release script size actually reduced ~800 bytes even though this PR is adding more logic.
-	Added a new file “imageRequest.test.ts”, in which first 211 lines are moved from “ajax.test.ts”, and added seven more UTs to cover everything in the new implementation.

File: src/ui/map.ts [hunk 2]
```diff
@@ -335,6 +340,9 @@ class Map extends Camera {
     _pixelRatio: number;
     _terrainDataCallback: (e: MapStyleDataEvent | MapSourceDataEvent) => void;
 
+    /** image queue throttling handle. To be used later when clean up */
+    _imgQHandle: number;
```
Comment (HarelM) [inline src/ui/map.ts:344]: Use full name: `imageQueueHandle`

File: src/ui/map.ts [hunk 3]
```diff
@@ -310,6 +314,7 @@ class Map extends Camera {
     _styleDirty: boolean;
     _sourcesDirty: boolean;
     _placementDirty: boolean;
+    _imageRequestQueueDirty: boolean;
```
Comment (HarelM) [inline src/ui/map.ts:317]: Shouldn't this be part of the queue?

File: src/util/imageRequest.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
+import type {Cancelable} from '../types/cancelable';
```
Comment (HarelM) [inline src/util/imageRequest.ts:1]: File name should match convention.

File: src/util/imageRequest.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+import type {RequestParameters, ExpiryData} from './ajax';
+import type {Callback} from '../types/callback';
+
+import {arrayBufferToImageBitmap, arrayBufferToImage} from './util';
+import {getArrayBuffer} from './ajax';
+import webpSupported from './webp_supported';
+import config from './config';
+
+export type GetImageCallback = (error?: Error | null, image?: HTMLImageElement | ImageBitmap | null, expiry?: ExpiryData | null) => void;
```
Comment (HarelM) [inline src/util/imageRequest.ts:10]: Probably not part of this PR, but I would like to move to promises.

File: src/util/imageRequest.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        imageRequestQueue.push(queued);
+
+        if (!isThrottled()) {
+            processQueue();
+        }
+
+        return queued;
+    }
+
+    const  arrayBufferToCanvasImageSource = (data: ArrayBuffer, callback: Callback<CanvasImageSource>) => {
```
Comment (HarelM) [inline src/util/imageRequest.ts:133]: Double space? Some are functions some are arrow functions?

**Comment by HarelM** [comment]
Overall this is alot better.
There are some nested callback hell which I hope @smellyshovel can solve after this gets merged.
I've added some minor comments.
Not sure I fully see the value of a namespace over a class, and since namespaces are not currently in use in this project I'm not sure this is the right place to start...
But overall, very nice work.

File: src/ui/map.ts (hunk 3 above)
Comment (zhangyiatmicrosoft) [reply src/ui/map.ts:317]: Not really. ImageRequest.processQueue returns the number of items remaining in the queue.
True of this flag means there are more too process in next frame and more render pass is needed.

File: src/util/imageRequest.ts (hunk 6 above)
Comment (zhangyiatmicrosoft) [reply src/util/imageRequest.ts:133]: Good catch.

File: src/util/imageRequest.ts (hunk 5 above)
Comment (zhangyiatmicrosoft) [reply src/util/imageRequest.ts:10]: Correct. I will defer this to #2038

**Comment by zhangyiatmicrosoft** [comment]
>.... Not sure I fully see the value of a namespace over a class

namespace generates an anonymous wrapper. Equivalent of a class with all static methods, and far more compact when it comes to minimized scripts.
I have seen other projects where "amd" is used in tsconfig which automatically wraps things of each file. It is certainly an option I can explore later for this project as well.
As it is now if not using namespace, everything would be just flat, which is not what you want either (from earlier iterations).

**Review by HarelM — APPROVED** [review]

**Comment by HarelM** [comment]
THANKS!
